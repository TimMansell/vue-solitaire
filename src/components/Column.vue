<template>
  <div
    class="column"
    @click="setColumn(columnNo)"
    @drop="dropCard(columnNo)"
    @dragover.prevent
    @dragenter.prevent
    :data-test="`column-${columnNo}`"
  >
    <Card
      v-for="(card, index) in cards"
      :key="index"
      :id="card.id"
      :value="card.value"
      :suit="card.suit"
      :order="card.order"
      :revealed="card.revealed"
      :visible="card.visible"
    />

    <SvgIcon
      class="card-placeholder"
      data-test="card-placeholder"
      v-if="!cards.length"
      name="Card_back_15"
    />
  </div>
</template>

<script>
import { mapGetters, mapActions } from 'vuex';
import Card from '@/components/Card.vue';
import SvgIcon from '@/components/SvgIcon.vue';

export default {
  name: 'Column',
  components: {
    Card,
    SvgIcon,
  },
  props: {
    cards: {
      type: Array,
      default: () => [],
    },
    columnNo: {
      type: Number,
      default: 0,
    },
  },
  computed: {
    ...mapGetters(['selectedCardId']),
  },
  methods: {
    ...mapActions(['moveCardsToColumn']),
    setColumn(columnNo) {
      const { selectedCardId } = this;

      if (selectedCardId) {
        this.moveCardsToColumn(columnNo);
      }
    },
    dropCard(columnNo) {
      this.moveCardsToColumn(columnNo);
    },
  },
};
</script>

<style scoped lang="scss">
.column {
  padding-left: $col-padding;
  padding-right: $col-padding;

  @media (min-width: $bp-md) {
    padding-left: $col-padding-lg;
    padding-right: $col-padding-lg;
  }
}

.card-placeholder {
  opacity: 0.1;
}
</style>
