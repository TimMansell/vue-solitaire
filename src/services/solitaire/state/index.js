const setState = (state, newState) => ({ ...state, ...newState });

export default setState;
