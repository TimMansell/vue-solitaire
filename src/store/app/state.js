// Return state as function so we can reset it.
const state = () => ({
  game: {
    moves: [],
    time: 0,
  },
  gameOutcome: {
    hasGameWon: false,
    hasGameLost: false,
  },
  isGameLoading: true,
  isGamePaused: false,
  isOverlayVisible: true,
  versionMatch: true,
});

export default state;
