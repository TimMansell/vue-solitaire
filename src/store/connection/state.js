const state = () => ({
  isConnecting: false,
  hasConnected: false,
  isOnline: false,
});

export default state;
