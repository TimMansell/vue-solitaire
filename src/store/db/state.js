// Return state as function so we can reset it.
const state = () => ({
  userStats: {
    totalGames: null,
  },
  game: {
    id: null,
    moves: 0,
  },
});

export default state;
