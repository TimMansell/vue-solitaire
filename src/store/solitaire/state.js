// Return state as function so we can reset it.
const state = () => ({
  board: {
    cards: [],
    foundation: [],
  },
  selectedCardId: null,
  isGameWon: false,
  hasMoves: true,
});

export default state;
