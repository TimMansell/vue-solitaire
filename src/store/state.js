const state = {
  cards: {
    values: ['A', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'J', 'Q', 'K'],
    suits: ['c', 's', 'h', 'd'],
  },
  shuffledCards: [],
  board: {
    cards: [],
    foundation: [],
  },
  rules: {
    columns: [7, 7, 7, 7, 6, 6, 6, 6],
    foundationColumns: [1, 1, 1, 1],
  },
  selectedCards: [],
};

export default state;
