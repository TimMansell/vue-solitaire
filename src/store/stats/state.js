// Return state as function so we can reset it.
const state = () => ({
  globalStats: [],
  userStats: [],
  playerCount: 0,
  onlinePlayerCount: 0,
  userGameCount: 0,
  globalGameCount: 0,
  leaderboards: [],
});

export default state;
