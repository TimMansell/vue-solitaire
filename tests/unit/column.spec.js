import { shallowMount, createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';
import Column from '@/components/Column.vue';
import state from '@/store/state';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('Column.vue', () => {
  let store;

  const getters = {
    selectedCardId: () => 1,
  };

  const actions = {
    moveCardsToColumn: jest.fn(),
  };

  beforeEach(() => {
    store = new Vuex.Store({
      state,
      getters,
      actions,
    });
  });

  it('matches snapshot', () => {
    const wrapper = shallowMount(Column, { store, localVue });

    expect(wrapper).toMatchSnapshot();
  });

  it('should have correct props', () => {
    const propsData = {
      cards: [
        {
          value: 'K',
          suit: 'd',
          order: 1,
          visible: true,
          clickable: false,
        },
      ],
      columnNo: 1,
    };

    const wrapper = shallowMount(Column, {
      store,
      localVue,
      propsData,
    });

    expect(wrapper.props().cards).toBe(propsData.cards);
    expect(wrapper.props().columnNo).toBe(1);
  });

  it('should call store action "moveCardsToColumn" when clicked', () => {
    const wrapper = shallowMount(Column, { store, localVue });

    wrapper.find('[data-test="column-0"]').trigger('click');

    expect(actions.moveCardsToColumn).toHaveBeenCalled();
  });
});
