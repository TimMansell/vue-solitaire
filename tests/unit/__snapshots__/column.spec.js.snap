// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column.vue matches snapshot 1`] = `<div data-test="column-0" class="column"></div>`;
